## Copyright 2021 Parker Dunn parker_dunn@outlook.com
  
#### Alternate: pgdunn@bu.edu & pdunn91@gmail.com   
#### July 20th, 2021

### Codecademy - Page Visits Funnel

Skill Path: Analyze Data with Python  
Section: Data Manipulation with Pandas  
Topic: Multiple tables in Pandas

### Assignment Context

This mini-project was meant to be a demonstration of what I have learned about working with multiple pandas dataframe tables. It is one part of a larger lesson about manipulating data with pandas/dataframes.  
  
This assignment is sort of a continuation of two previous pandas projects: "Petal Power Inventory" and "A/B Testing for ShoeFly." The collection of three projects demonstrates many of the things that I have learned to do with the pandas library.

### Assignment Description

"Cool T-Shirts Inc. has asked you to analyze data on visits to their website. Your job is to build a *funnel*, which is a description of how many people continue to the next step of a multi-step process.  
  
In this case, our funnel is going to describe the following process:
1. A user visits CoolTShirts.com
2. A user adds a t-shirt to their cart
3. A user clicks "checkout"
4. A user actually purchases a t-shirt"

In [1]:
# import codecademylib
# Data provided for the assignment in visits.csv, cart.csv, checkout.csv, and purchase.csv

"*import codecademylib*" was included the assignment template. This is not a package that can be imported here so that line is commented out.

The contents of *codecademylib* were not explicitly provided. The webpage and contents can be downloaded but the contents of the package are not clearly specified.

I was able to open and generate a copy of *visits.csv*, *cart.csv*, *checkout.csv*, and *purchase.csv* so that the files can still be imported and the script can be run here.

In [2]:
import pandas as pd

visits = pd.read_csv("visits.csv", parse_dates = [1])
cart = pd.read_csv("cart.csv", parse_dates = [1])
checkout = pd.read_csv("checkout.csv", parse_dates = [1])
purchase = pd.read_csv("purchase.csv", parse_dates = [1])

print("Visits - First 8 rows\n",visits.head(8))
print("\nCart - first 8 rows\n", cart.head(8))
print("\nCheckout - first 8 rows\n",checkout.head(8))
print("\nPurchase - first 8 rows \n", purchase.head(8))

Visits - First 8 rows
                                 user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00
6  6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00
7  49a6399b-c095-4e42-92eb-af90abe2bb6a 2017-01-10 12:09:00

Cart - first 8 rows
                                 user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
5  0c998868-9961-490b-8ec2-9981ba667027 2017-11-27 08:5

In [3]:
# Requested to combine visits and cart using a left merge
# Order not specified but seems like we are looking at all visits then ...
# how many visits turned into users starting a cart

visits_cart = pd.merge(visits, cart, how="left")
print(visits_cart.head(5), visits_cart.tail(5))
print("\nNumber of entries in merged df: {}\n".format(visits_cart.user_id.count()))
print("\n",len(visits_cart))   # Alternate approach suggested

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT                                      user_id          visit_time cart_time
1995  33913ac2-03da-45ae-8fc3-fea39df827c6 2017-03-25 03:29:00       NaT
1996  4f850132-b99d-4623-80e6-6e61d003577e 2017-01-08 09:57:00       NaT
1997  f0830b9b-1f5c-4e74-b63d-3f847cc6ce70 2017-09-07 12:56:00       NaT
1998  b01bffa7-63ba-4cd3-9d93-eb1477c23831 2017-07-20 04:37:00       NaT
1999  0336ca81-8d68-443f-9248-ac0b8ad147d5 2017-11-15 10:11:00       NaT

Number of entries in merged df: 2000



In [4]:
# Requested: "How many timestamps are null for the cart_time col?"

print("Number of null timestamps: {}"\
          .format(visits_cart.cart_time.isnull().sum()))

Number of null timestamps: 1652


The fact that there are 1652 null timestamps for "cart_time" in the merged dataframe means 1652 occurances of someone visiting the company website and not adding anything to the cart.  
  
This also suggests that 348 of 2000 visitors to the website added a t-shirt to their cart.

__Checking on some information about the length and contents of the dataframes__

In [5]:
# Visits.csv

print("There are {} entries in the visits dataframe".format(len(visits)))
print("\nThere are {} unique user IDs in the visits dataframe.".format(visits.user_id.nunique()))

There are 2000 entries in the visits dataframe

There are 2000 unique user IDs in the visits dataframe.


In [6]:
# cart.csv

print("There are {} entries in the cart dataframe.".format(len(cart)))
print("\nThere are {} unique user IDs in the cart dataframe.".format(cart.user_id.nunique()))

There are 348 entries in the cart dataframe.

There are 348 unique user IDs in the cart dataframe.


In [7]:
# checkout.csv

print("There are {} entries in the checkout dataframe.".format(len(checkout)))
print("\nThere are {} unique user IDs in the checkout dataframe.".format(checkout.user_id.nunique()))

There are 360 entries in the checkout dataframe.

There are 226 unique user IDs in the checkout dataframe.


The instructions in the assignment do indicate that I was supposed to account for the repeated unique user IDs in checkout.  

From the dataframe information printed in the three cells above, there is clearly repeated user IDs, which isn't necessarily a barrier to working with the data. However, later in the assignment (*see below*), I was instructed to:  

"Repeat the left merge for `cart` and `checkout` and count `null` values. What percentage of users put items in their cart, but did not proceed to checkout?"  

Since the assignment did not indicate at all that I should account for repeat user IDs in the `checkout` data, I will not change any of the calculations below. However, The next section will previous extract this checkout data and show some of the repeats for confirmation.


In [33]:
# Grouping checkout by user_id and displaying frequency of user_ids

checkout_by_user = checkout.groupby("user_id").checkout_time.count().reset_index()
print(checkout_by_user.head(20))
print(type(checkout_by_user))

                                 user_id  checkout_time
0   00065d4e-909c-48b5-a69e-27a1abbdbb87              1
1   009270e0-1d69-4090-9f9a-b0b874b7e6e0              1
2   0182a7b9-2099-4476-8f7d-51f2c2f71dff              2
3   02e684ae-a448-408f-a9ff-dcb4a5c99aac              2
4   04eaa41a-e9d0-4d10-a01b-8f21ec2337cd              1
5   05b44764-bb83-4b08-b3ff-c6b31d4e31d3              1
6   079a26c7-e8e8-4ac5-bd92-1ec65edb1830              1
7   0ae3bec2-a3b2-4803-9a00-7aa64ebb27f9              1
8   0b69d940-dd3c-4851-88dd-8044f4be13d2              1
9   0c230d3b-3da4-47f6-860f-71e2bc4030b9              1
10  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2              1
11  0c998868-9961-490b-8ec2-9981ba667027              1
12  0d47febd-a7dc-487f-a65a-698c62ab8696              1
13  0d7dd72a-4671-4d44-abaa-45497615bfcd              2
14  0ea4cc68-dae4-4e35-b3e0-f0889932e1b5              2
15  0f9c4be6-1b4f-4500-aebc-73e46a09c52e              1
16  0fe7eefc-a294-44c3-b462-d78ae2016ccc        

In [19]:
# Requested: "What percent of users who visited Cool T-Shirts Inc. ended up not 
# placing a t-shirt in their cart?"

per_no_cart = (float(visits_cart.cart_time.isnull().sum()) / len(visits_cart))

print("Percentage of users who did not add a t-shirt to their cart: {:.2%}"\
                .format(per_no_cart))

Percentage of users who did not add a t-shirt to their cart: 82.60%


In [17]:
# Asked to complete a similar "left" merge again with cart and checkout this time
# counting null values again and determining percentage of users
# who did not go to checkout

cart_checkout = pd.merge(cart, checkout, how="left")

# len(cart_checkout) should contain all entries from cart
# regardless of if there is a matching entry in checkout

checkout_time_na = cart_checkout.checkout_time.isnull().sum()

per_no_checkout = float(checkout_time_na) / len(cart_checkout)
print(len(cart_checkout))
print("Null time values for 'checkout_time': {}".format(checkout_time_na))
print("Percentage of users who did not checkout: {:.2%}"\
          .format(per_no_checkout))

482
Null time values for 'checkout_time': 122
Percentage of users who did not checkout: 25.31%


*__I do not see a problem with my work at this moment, but the numbers don't seem to add up here.__*  


It seems odd that we could have 348 users who visited the site and placed an item in their cart then end up with 482 users (a greater number) who had an item in their cart.  
  
Unless I'm missing something right now, the left merge that creates "cart_checkout" should not have more values than the number of carts we found were created when analyzing the "visits_cart" data frame.

_____

Update: I added sections above demonstrating the lengths of the original dataframes. Based on those original data frames, I am not sure how the dataframe merge created a dataframe with more entries than either of the original data frames. I believe the use of a "left" merge should have included all `user_id` entries from *cart* and only matching rows from *checkout*.  
  
I will further investigate this issue if I have time at some point. However, I am not particularly interested in this particular subject and believe I understand how to use pandas the way that this lesson was trying to teach.  
  
Using a different different variation of `merge()` in the next section to double check that the "left" merge happened as expected.


In [21]:
# Checking on the cart and checkout merge

cart_checkout_outer = pd.merge(cart, checkout, how="outer")
print("Size of outer merge: {}".format(len(cart_checkout_outer)))

cart_checkout_specific_cols = pd.merge(cart, checkout, left_on="user_id",\
                                      right_on="user_id", how="left")
print("Size of left merge with specific cols: {}".format(len(cart_checkout_specific_cols)))

Size of outer merge: 482
Size of left merge with specific cols: 482


**well..... the left merge appears to be performing the same way as an outer merge .....** 

I'm not sure why the "left" merge doesn't seem to be working correctly.  
  
Based on the documentation on `.merge()`: "left: use only keys from the left frame"

In [24]:
# Checking on inner merge of cart and checkout

print("Unique entries in cart: {}".format(cart.user_id.nunique()))
print("Unique entries in checkout: {}".format(checkout.user_id.nunique()))

cart_checkout_inner = pd.merge(cart, checkout, how="inner")
print("\nSize of inner merge: {}".format(len(cart_checkout_inner)))

Unique entries in cart: 348
Unique entries in checkout: 226

Size of inner merge: 360


Looks like there are two issues going on here.  
1. There are multiple entries for some user_ids in checkout - which was briefly mentioned above
2. It seems that if you only include user_ids from cart, then there are 360 entries for these user_ids
3. I believe there are many entries in checkout for user_ids that do not match those of the user_ids in cart. Maybe I'm misunderstanding the user_id system but this seems weird to me.
  
To confirm the presence of new user_ids in the checkout, I'm going to do an explicit `for` loop to check on this fact.

In [26]:
# First checking out the list/series of user_ids from cart

user_ids_cart = cart.user_id.to_list()
print(user_ids_cart[:9])
print(type(user_ids_cart))

['2be90e7c-9cca-44e0-bcc5-124b945ff168', '4397f73f-1da3-4ab3-91af-762792e25973', 'a9db3d4b-0a0a-4398-a55a-ebb2c7adf663', 'b594862a-36c5-47d5-b818-6e9512b939b3', 'a68a16e2-94f0-4ce8-8ce3-784af0bbb974', '0c998868-9961-490b-8ec2-9981ba667027', 'f783c680-1d9a-437d-9f45-7827299b78fa', 'f3dd7631-3cf6-4d49-9c6e-b48eda4432b5', '4cdd357a-199e-4d81-ae9e-7252fa34ee0f']
<class 'list'>


In [ ]:
# Now interating through the list of user_ids from checkout

user_ids_checkout = checkout.user_id.to_list()

counter = 0

for id in user_ids_checkout:
    if (id in user_ids_cart):
        continue
    else:
        counter+=1

print(counter)

## Well counter ends up being 0 .... soooo where are the extra entries coming from ....

# Grouped checkout in an above section -- borrowring that df here
print("Total count of the grouped checkout entries: {}".format(checkout_by_user.checkout_time.sum()))

I cannot make sense of the values that I am getting for the checkout and cart merge.
* Number of entries for left merge: 482
* Number of entries for the outer merge: 482 - this should be the theoretical max
* Number of entries for inner merge: 360
  
* Number of user_ids in checkout but not in cart: 0

*NOTE: The number above would have been what I expected in practice but does not match the other numbers as far as I can tell so far.*
  
* Number of unique entries in cart: 348
* Number of unique entries in checkout: 226  
  
For the left merge, there were 122 null time values for "checkout_time."

### AHHHHH okay I figured it out

There are 348 user_ids in cart.  
These user_ids are the only ones associated with the user_ids in checkout.  
There are 360 instances of a user_id from cart clicking on checkout.  
There are 122 instances of a user_id from cart not clicking on checkout.  
  
  
Then, of the 348 users from cart if we take out the 122 that did not select checkout we end up with the 226 unique users from checkout!

In [10]:
# asked to combine all of the data now using a series of left merges

all_data = visits.merge(cart, how="left")\
                 .merge(checkout, how="left")\
                 .merge(purchase, how="left")
print(all_data.head(10))

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
5  7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00   
6  6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00   
7  49a6399b-c095-4e42-92eb-af90abe2bb6a 2017-01-10 12:09:00   
8  653c095d-fd74-40dd-ae3b-5a69a2613629 2017-06-27 19:21:00   
9  3d9fdac1-aae3-4a01-b448-934d12047468 2017-09-24 06:43:00   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  
2 2017-08-20 08:31:00                 NaT                 NaT  
3                 NaT                 NaT         

In [11]:
# Asked to find "What percentage of users proceeded to checkout, but did not purchase a t-shirt?"

num_checkout = len(all_data) - all_data.checkout_time.isnull().sum()
num_purchase = len(all_data) - all_data.purchase_time.isnull().sum()
checkout_no_purchase = (num_checkout - num_purchase)
print("\nNumber of Checkouts: {} | Number of Purchases: {}\n"\
          .format(num_checkout, num_purchase))
print("Number of users who started checkout and did not make a purchase: {}"\
          .format(checkout_no_purchase))

print("\nPercentage of users who proceeded to checkout but did not make a purchase: {:4%}"\
                     .format(float(checkout_no_purchase)/num_checkout))


Number of Checkouts: 598 | Number of Purchases: 497

Number of users who started checkout and did not make a purchase: 101

Percentage of users who proceeded to checkout but did not make a purchase: 16.889632%


In [35]:
# Confirming there are no entries that ...
# have a purchase_time without a checkout_time

check_null = lambda purchase_time: True if (purchase_time == "NaT") else False

# Checking on what "all_data.checkout_time.isnull()" looks like
print(all_data.checkout_time.isnull().head(10))

0     True
1    False
2     True
3     True
4     True
5     True
6     True
7     True
8     True
9     True
Name: checkout_time, dtype: bool


In [37]:
checkout_null = all_data[all_data.checkout_time.isnull()] # contains only entries with no "checkout_time"


# checkout_null["purchase_null"] = checkout_null.purchase_time.apply(check_null)
# print(checkout_null.head(10))
# purchase_without_checkout = checkout_null.groupby("purchase_null").count()

print("If the numbers below are the same size, then all null checkout_times have null purchase_times")
print(len(checkout_null))
print(len(checkout_null.purchase_time.isnull()))

If the numbers below are the same size, then all null checkout_times have null purchase_times
1774
1774


In summary...  
* 16.89% of people __did checkout__ but __did not purchase__
* 25.31% of users __did add an item to cart__ but __did not proceed to checkout__
* 82.6% of users __visited site__ but __did not add item to cart__

The weakest step of the funnel is getting users to add an item to their cart once they are on the site.  
  
As noted above, these are the numbers that were requested (as far as I could tell). However, these numbers do not reflect the fact that some user_ids appear in checkout multiple times. Thus, the 25.31% of users who "added an item to their cart but did not proceed to checkout" is actually higher than it looks.  

In reality, 122 users from cart did not select checkout from the total of 348 users in cart. This would put the percentage of people who added an item to cart and did not checkout at... (see below)

In [41]:
print("Percent of users in cart who did not checkout (using number of unique users in cart): {:.2%}"\
                         .format(122.0/348))

Percent of users in cart who did not checkout (using number of unique users in cart): 35.06%


In [43]:
# Requested to provide the average time from initial visit to final purchase

all_data['time_to_purchase'] = all_data.purchase_time \
                                    - all_data.visit_time

print(all_data.time_to_purchase)
print(all_data.time_to_purchase.mean())

0                  NaT
1      0 days 00:44:00
2                  NaT
3                  NaT
4                  NaT
             ...      
2367               NaT
2368               NaT
2369               NaT
2370               NaT
2371               NaT
Name: time_to_purchase, Length: 2372, dtype: timedelta64[ns]
0 days 00:43:53.360160965


Average time to purchase was about 44 minutes.